### PR TITLE
Hybrid Parallel (AD): Fix Race Conditions

### PR DESCRIPTION
**Pull request by jblueh**
## Proposed Changes
@maxaehle and I analyzed SU2 with `gcc` with the thread sanitizer enabled and fixed various race conditions both in hybrid parallel and hybrid parallel AD code.

The thread sanitizer records of the hybrid regression tests and hybrid regression AD tests are clean now.

## Related Work
#1384, this PR works towards fixing it

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [x] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: Common/src/geometry/CGeometry.cpp [hunk 1]
```diff
@@ -2506,38 +2506,41 @@ void CGeometry::UpdateCustomBoundaryConditions(CGeometry **geometry_container, C
 }
 
 void CGeometry::ComputeSurfaceAreaCfgFile(const CConfig *config) {
-  const auto nMarker_Global = config->GetnMarker_CfgFile();
-  SurfaceAreaCfgFile.resize(nMarker_Global);
-  vector<su2double> LocalSurfaceArea(nMarker_Global, 0.0);
+  SU2_OMP_MASTER
+  {
+    const auto nMarker_Global = config->GetnMarker_CfgFile();
```
[the remark this answers is not in the document]
Comment (jblueh) [reply Common/src/geometry/CGeometry.cpp:2511]: How would we safeguard the `SU2_MPI::Allreduce` at the end of `CGeometry::ComputeSurfaceAreaCfgFile` for the case of funneled comm mode?

File: TestCases/TestCase.py [hunk 2]
```diff
@@ -695,7 +708,7 @@ def adjust_iter(self):
         # Rewrite the file with a .autotest extension
         self.cfg_file = "%s.autotest"%self.cfg_file
         file_out = open(self.cfg_file,'w')
-        file_out.write('%% This file automatically generated by the regression script\n')
+        file_out.write('% This file automatically generated by the regression script\n')
         file_out.write('%% Number of iterations changed to %d\n'%(self.test_iter+1))
```
[the remark this answers is not in the document]
Comment (jblueh) [reply TestCases/TestCase.py:712]: The `%%` is needed to escape the modulo formatting and actually print the `%`.

File: SU2_CFD/src/solvers/CDiscAdjSolver.cpp [hunk 3]
```diff
@@ -584,7 +584,9 @@ void CDiscAdjSolver::SetSurface_Sensitivity(CGeometry *geometry, CConfig *config
 
 void CDiscAdjSolver::Preprocessing(CGeometry *geometry, CSolver **solver_container, CConfig *config, unsigned short iMesh,
                                    unsigned short iRKStep, unsigned short RunTime_EqSystem, bool Output) {
+  SU2_OMP_MASTER
   config->SetGlobalParam(config->GetKind_Solver(), RunTime_EqSystem);
+  END_SU2_OMP_MASTER
```
[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/solvers/CDiscAdjSolver.cpp:589]: In the hybrid parallel AD tests, it is only reported as a write-write race from within a parallel region where no additional read occurs. Are you aware of another call tree that would require barriers? In general, we could of course default to a master preceeded and/or followed by barrier pattern, like the `SU2_OMP_SAFE_GLOBAL_ACCESS` you suggested below.

[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/solvers/CDiscAdjSolver.cpp:589]: Got it, there is probably a barrier downstream, I'm in favor of a "standard" solution, like the proposed macro, makes it easier to recommend best practices.

File: Common/src/linear_algebra/CSysMatrix.cpp [hunk 4]
```diff
@@ -641,7 +641,7 @@ template<class ScalarType>
 void CSysMatrix<ScalarType>::BuildJacobiPreconditioner() {
 
   /*--- Build Jacobi preconditioner (M = D), compute and store the inverses of the diagonal blocks. ---*/
-  SU2_OMP_FOR_(schedule(dynamic,omp_heavy_size) SU2_NOWAIT)
+  SU2_OMP_FOR_(schedule(dynamic,omp_heavy_size) )
```
[the remark this answers is not in the document]
Comment (jblueh) [reply Common/src/linear_algebra/CSysMatrix.cpp:644]: This change and the one in 7cffa8e85a72264640783f0050e40e4925453ac6 fix thread sanitizer findings that point neither to SU2 nor the AD tools but `libgomp`. Those could also be false positives.

File: TestCases/TestCase.py [hunk 5]
```diff
@@ -760,3 +773,42 @@ def disable_restart(self):
         os.chdir(workdir)
 
         return
+
+    def disable_preacc(self):
+
+        # Read the cfg file
```
[the remark this answers is not in the document]
Comment (jblueh) [reply TestCases/TestCase.py]: Reverted in 0b021a73e50ef17e5fdf90889ab24a2ba205767e

File: SU2_CFD/src/integration/CMultiGridIntegration.cpp [hunk 6]
```diff
@@ -226,7 +226,13 @@ void CMultiGridIntegration::MultiGrid_Cycle(CGeometry ****geometry,
 
     /*--- Temporarily disable implicit integration, for what follows we do not need the Jacobian. ---*/
 
-    if (implicit) config->SetKind_TimeIntScheme(EULER_EXPLICIT);
+    if (implicit) {
+      SU2_OMP_BARRIER
+      SU2_OMP_MASTER
+        config->SetKind_TimeIntScheme(EULER_EXPLICIT);
+      END_SU2_OMP_MASTER
+      SU2_OMP_BARRIER
```
[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/integration/CMultiGridIntegration.cpp]: I added two variants of this in 8ed32fcdf6e491a7902876e6fd54e25c19f73750,  the suggested one and a `BEGIN_SU2_OMP_SAFE_GLOBAL_ACCESS` and `END_SU2_OMP_SAFE_GLOBAL_ACCESS` pair that feels less awkward for protecting scopes of multiline code. As a bonus, the syntax checker detects incorrect pairing.

I applied this throughout the code for all occurrencies of `barrier` + `master` + `barrier`. There are several code locations where only one of the barriers is present or none, I did not touch these yet.

File: Common/src/linear_algebra/CSysMatrix.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jblueh) [reply Common/src/linear_algebra/CSysMatrix.cpp:644]: As discussed in the dev meeting, we keep these changes for the sake of a clean thread sanitizer record.

File: SU2_CFD/src/integration/CMultiGridIntegration.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/integration/CMultiGridIntegration.cpp]: Further applications in 7a2eb435a2e477ed8b0059e590ea96cfe0b5b252, as discussed in the dev meeting.

File: Common/include/parallelization/vectorization.hpp [hunk 7]
```diff
@@ -56,8 +56,18 @@ constexpr size_t PREFERRED_SIZE = 8;
  */
 template<class T>
 constexpr size_t preferredLen() { return PREFERRED_SIZE / sizeof(T); }
+
+#if !defined(CODI_REVERSE_TYPE) && !defined(CODI_FORWARD_TYPE)
 template<>
 constexpr size_t preferredLen<su2double>() { return PREFERRED_SIZE / sizeof(passivedouble); }
+#else
+/* enforce one element per SIMD array since (1) arrays of AD types cannot be auto vectorized and (2) they might violate
+ * exclusive read declarations if masking is not applied throughout */
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply Common/include/parallelization/vectorization.hpp]: Makes sense. If I understood the problem correctly there is a better-performing fix by passing the "Nodes" array.
If not please revert https://github.com/su2code/SU2/pull/1679/commits/cf1a98d202364666bdd3f2ad830a60894bcb7bc3

[the remark this answers is not in the document]
Comment (jblueh) [reply Common/include/parallelization/vectorization.hpp]: LGTM :+1: I also rechecked the hybrid parallel and hybrid parallel AD tests with the thread sanitizer and nothing came up.

**Review by pcarruscag — APPROVED** [review]

File: Common/include/parallelization/vectorization.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (pcarruscag) [reply Common/include/parallelization/vectorization.hpp]: Great, thanks for checking.

**Pull request merged**